File: coding-challenges/sources/daily-coding-problem/Day 20/Day20.swift


fileprivate final class Node {
    var val: Int
    var next: Node?

    init(_ val: Int, next: Node? = nil) {
        self.val = val
        self.next = next
    }
}

extension Node {
    var length: Int {
        // Return 0 if the list is empty
        guard let next = self.next else { return 0 }

        // Return 1 plush length of the rest of the list
        return 1 + next.length
    }
}

fileprivate func intersection(_ a: Node?, _ b: Node?) -> Node? {
    // Determine the length of each linked list
    let m = a?.length ?? 0
    let n = b?.length ?? 0

    // Initialize pointers to the head of each list
    var curA = a
    var curB = b

    // Advance the pointer for the longer list by the difference in lengths
    if m > n {
        var steps = m - n
        repeat {
            curA = curA?.next
            steps -= 1
        } while steps > 0
    } else {
        var steps = m - n
        repeat {
            curB = curB?.next
            steps -= 1
        } while steps > 0

    }

    // Iterate through both lists in parallel, comparing the nodes at each step
    while curA !== curB {
        curA = curA?.next
        if curA === curB {
            return curA
        }
        curB = curB?.next
    }

    // Return the node where the pointers intersect
    return curA
}

import XCTest

final class Day20: XCTestCase {
    func test_intersectionAtBeginning() {
        let intersectingNode = Node(2, next: Node(3))
        let a = Node(1, next: intersectingNode)
        let b = intersectingNode
        XCTAssertEqual(intersection(a, b)?.val, 2)
    }

    func test_intersectionAtEnd() {
        let intersectingNode = Node(3)
        let a = Node(1, next: Node(2, next: intersectingNode))
        let b = Node(4, next: Node(5, next: intersectingNode))
        XCTAssertEqual(intersection(a, b)?.val, 3)
    }

    func test_noIntersection() {
        let a = Node(1, next: Node(2, next: Node(3)))
        let b = Node(4, next: Node(5, next: Node(6)))
        XCTAssertNil(intersection(a, b)?.val)
    }
}

File: coding-challenges/sources/daily-coding-problem/Day 20/Day20.kt
fun main(args: Array<String>) {
    findIntersectionTests()
}

class Node(val value: Int, var next: Node? = null)

fun Node?.length(): Int {
    // Return 0 if the list is empty
    if (this == null) {
        return 0
    }
    // Return 1 plus the length of the rest of the list
    return 1 + (this.next?.length() ?: 0)
}


fun findIntersection(a: Node? , b: Node?): Node? {
    // Determine the lenght of the each linked list
    val m = a?.length() ?: 0
    val n = b?.length() ?: 0
    
    // Pointers to the head of each list
    var curA = a
    var curB = b
    
    // Advaince the pointer for the longer list by the difference in lengths
    if (m > n) {
        repeat(m - n) {
            curA = curA?.next
        }
    } else {
        repeat(n - m) {
            curB = curB?.next
        }
    }
    
    // Iterate through both lists in parallel, comparing the nodes at each step
    while (curA !== curB) {
        curA = curA?.next
        if (curA == curB) return curA
        curB = curB?.next
    }
    
    return curA
}

fun findIntersectionTests() {
    fun testIntersectionAtBeginning() {
        val intersectingNode = Node(2, Node(3))
        val a = Node(1, intersectingNode)
        val b = intersectingNode
        println(findIntersection(a, b) == intersectingNode)
    }
    
    fun testIntersectionAtEnd() {
        val intersectingNode = Node(3)
        val a = Node(1, Node(2, intersectingNode))
        val b = Node(4, Node(5, intersectingNode))
        println(findIntersection(a, b) == intersectingNode)
    }
    
    fun testNoIntersection() {
        val a = Node(1, Node(2, Node(3)))
        val b = Node(4, Node(5, Node(6)))
        println(findIntersection(a, b) == null)
    }
    
    testIntersectionAtBeginning()
    testIntersectionAtEnd()
    testNoIntersection()
}
